.github/workflows/publish-release.yml

```diff
@@ -0,0 +1,229 @@
+          version=$(grep --max-count=1 "<version>.*</version>" pom.xml | sed -E -e "s/\s*<[^>]+>//g" -e "s/-SNAPSHOT//" -e "s/-[0-9.]+//g")
+
+          # update change
+          echo "Releasing ${#changes[@]} changes as version $version:"
+          for (( i=0; i<${#changes[@]}; i++ )); do echo "${changes[$i]}" ; done
+          sed -i "s/## \[UNRELEASED\] - YYYY-MM-DD/## [$version] - $(date +%Y-%m-%d)/" CHANGELOG.md
+          sed -i -e "s/$latest-/$version-/g" -e "s/$latest\./$version./g" README.md PYSPARK-DEPS.md python/README.md
+          ./set-version.sh $version
+
+          # commit changes to local repo
+          git config --global user.name "$(git --no-pager log --format=format:'%an' -n 1)"
+          git config --global user.email "$(git --no-pager log --format=format:'%ae' -n 1)"
+          echo "Committing release to local git"
+          git add pom.xml python/setup.py CHANGELOG.md README.md python/README.md
+          git commit -m "Releasing $version"
+          git tag -a "v${version}" -m "Release v${version}"
+
+          # updating master and pushing a tag for the release
+          echo "Pushing release commit and tag to origin"
+          git push origin master "v${version}"
+
+          # share release version to next job
+          echo "releaseTagVersion=v${version}" >> "$GITHUB_OUTPUT"
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+  publish-release:
+    name: Publish release
+    runs-on: ubuntu-latest
+    needs: [prepare-release]
+    permissions:
+      contents: write # required to push to a branch
+    strategy:
+        fail-fast: false
+        matrix: ${{fromJson(github.event.inputs.versions)}}
+
+    steps:
+      - name: Checkout release tag
+        uses: actions/checkout@v4
+        with:
+          ref: ${{needs.prepare-release.outputs.RELEASE_TAG}}
+
+      - name: Set up JDK 8
+        uses: actions/setup-java@v4
+        with:
+          java-version: '8'
+          distribution: 'corretto'
+
+      - uses: actions/setup-python@v5 
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+
+      - name: Cache Maven packages
+        id: cache-maven
+        uses: actions/cache@v4
+        with:
+          path: ~/.m2/repository
+          key: ${{ runner.os }}-mvn-build-${{ matrix.params.spark-version }}-${{ matrix.params.scala-version }}-${{ hashFiles('pom.xml') }}
+          restore-keys: ${{ runner.os }}-mvn-build-${{ matrix.params.spark-version }}-${{ matrix.params.scala-version }}-
+      
+      - name: Cache Pip packages
+        id: cache-pip
+        uses: actions/cache@v4
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-whl-${{ env.PYTHON_VERSION }}-${{ matrix.params.spark-version }}
+
+      - name: Import GPG key
+        id: sign
+        uses: crazy-max/ghaction-import-gpg@01dd5d3ca463c7f10f7f4f7b4f177225ac661ee4
```

Comment:
I think we need to set the `GPG_PRIVATE_KEY` secret. Which key do you propose we use?

Reply:
shouldn't we use a private key owned by GR here, not my personal one? Same with the OSSRH login, that should be a general GR login used by all packages that deploy to maven, not a personal access.

Reply:
I can provide the key and credentials as environment secrets

Reply:
That would be awesome!

Reply:
I think we also need PyPi credentials.

Reply:
`GPG_PRIVATE_KEY` and `GPG_PASSPHRASE` have been added as secrets under the `release` environment

Reply:
for PyPI publishing, the package should be transferred to the [`gresearch`](https://pypi.org/user/gresearch/) PyPI user and [trusted publishing](https://docs.pypi.org/trusted-publishers/) should be setup.

Reply:
`OSSRH_USERNAME` and `OSSRH_PASSWORD` have also been added as secrets under the `release` environment for Maven publishing.